Round geometry and bbox coordinates to 6 decimal places

pjhartzell opened the pull request:
**Related Issue(s):**
#44 

**Description:**
Added utility function to round Item geometry and bbox coordinates to 6 decimal places.

**PR checklist:**

- [x] Code is formatted (run `scripts/format`).
- [x] Code lints properly (run `scripts/lint`).
- [x] Tests pass (run `scripts/test`).
- [ ] Documentation has been updated to reflect changes, if applicable.
- [x] Examples have been updated to reflect changes, if applicable
- [x] Changes are added to the [CHANGELOG](../CHANGELOG.md).

pjhartzell committed:
rounds geometry and bbox coordinates to 6 decimal places

pjhartzell committed:
update changelog

pjhartzell committed:
rebase conflict fixes

pjhartzell committed:
add test

pjhartzell committed:
fix mypy

gadomski commented on examples/landsat-c2-l1/LM01_L1GS_001010_19720908_02_T2/LM01_L1GS_001010_19720908_02_T2.json:
```diff
@@ -7,7 +7,7 @@
       "mss"
     ],
-    "created": "2022-03-31T18:24:59.855547Z",
+    "created": "2022-10-15T12:22:11.833165Z",
```
IMO not worth it for this PR, but I'm starting to think we _shouldn't_ include `created` in our example files, in order to cut down on diff noise. IMO `stac.create_item` should still set `created`, but we should either
- strip it off in a `create_examples`script, or
- provide a command-line flag to not include the `created` field

I'm dropping this note here because I've been thinking about it but haven't thought about it during a PR review until now.

gadomski commented on src/stactools/landsat/constants.py:
```diff
@@ -51,3 +51,5 @@ class Sensor(Enum):
     },
 }
+
+COORD_PRECISION = 6
```
```suggestion
COORDINATE_PRECISION = 6
```

Nitpicky, but IMO no reason to abbreviate unless we already use "coord" as an abbreviation in this library.

pjhartzell replied:
NP. Done.

gadomski commented on src/stactools/landsat/utils.py:
```diff
@@ -98,3 +98,34 @@ def handle_antimeridian(item: Item, antimeridian_strategy: Strategy) -> None:
     antimeridian.fix_item(item, antimeridian_strategy)
+
+
+def round_coordinates(item: Item, precision: int = 6) -> None:
```
```suggestion
def round_coordinates(item: Item, precision: int = 6) -> Item:
```

It's undocumented, but I've started preferring a "fluent" interface where we both modify the item in place and return the modified item. This allows the function to be used in assignment-style setups (`item = round_coordinates(create_item("file.tif"))`) as well as mutation setups (`round-coordinates(item)`).

As a general note, this function does seem useful enough to push into **stactools** at a later date.

pjhartzell replied:
Yeah, we've discussed this concept. I did not include it in order to be consistent with the `handle_antimeridian` function. But I'm fine with changing it. I updated both `round_coordinates` and `handle_antimeridian` to be fluent, for consistency.

gadomski commented on src/stactools/landsat/utils.py:
```diff
@@ -98,3 +98,34 @@ def handle_antimeridian(item: Item, antimeridian_strategy: Strategy) -> None:
+    Args:
+        item (Item): pystac Item.
+        precision (int): Number of decimal places for rounding. Defaults to 6,
+            which is approximately 1 cm at the equator.
```
```suggestion
            which is approximately 1 cm at the equator.
        Returns:
            Item: The original PySTAC Item, with rounded coordinates.
```

gadomski commented on src/stactools/landsat/utils.py:
```diff
@@ -98,3 +98,34 @@ def handle_antimeridian(item: Item, antimeridian_strategy: Strategy) -> None:
+            which is approximately 1 cm at the equator.
+    """
+
+    def recursive_round(coordinates: List[Any], precision: int) -> List[Any]:
```
```suggestion
    def recursive_round(coordinates: Iterable[Any], precision: int) -> List[float]:
```

pjhartzell replied:
- `coordinates` actually needs to be a `List`, not a generic `Iterable`. For example, it cannot be a Tuple (that's pronounced _tupple_, mind you), because it would not be able to be updated in place. Note the conversion to list for both geometry and bbox before the call.
- The return type can be a List, or a few different structures of nested Lists, so `List[Any]` is more appropriate, no?

gadomski replied:
> coordinates actually needs to be a List, not a generic Iterable. For example, it cannot be a Tuple (that's pronounced tupple, mind you), because it would not be able to be updated in place. Note the conversion to list for both geometry and bbox before the call.

Ah yeah, I guess I didn't think about the implications of the in-place modifications. Disregard.

>The return type can be a List, or a few different structures of nested Lists, so List[Any] is more appropriate, no?

Correct, my mistake, disregard.

gadomski commented on src/stactools/landsat/utils.py:
```diff
@@ -98,3 +98,34 @@ def handle_antimeridian(item: Item, antimeridian_strategy: Strategy) -> None:
+    """
+
+    def recursive_round(coordinates: List[Any], precision: int) -> List[Any]:
+        for idx, _item in enumerate(coordinates):
```
Could we use `value` or something other than `_item` for this variable?

pjhartzell replied:
Yeah, `_item` is craptastically bad. Changed to `value`.

gadomski commented on src/stactools/landsat/utils.py:
```diff
@@ -98,3 +98,34 @@ def handle_antimeridian(item: Item, antimeridian_strategy: Strategy) -> None:
+
+    def recursive_round(coordinates: List[Any], precision: int) -> List[Any]:
+        for idx, _item in enumerate(coordinates):
+            if isinstance(_item, (list, tuple)):
+                coordinates[idx] = list(_item)  # handle any tuples
+                coordinates[idx] = recursive_round(coordinates[idx], precision)
+            elif isinstance(_item, (int, float)):
+                coordinates[idx] = round(_item, precision)
+            else:
+                raise ValueError(f"Unexpected type: {_item} encountered.")
```
```suggestion
            if isinstance(_item, (int, float)):
                coordinates[idx] = round(_item, precision)
            else:
                coordinates[idx] = list(_item)  # handle any tuples
                coordinates[idx] = recursive_round(coordinates[idx], precision)
```

I think we can just round numbers first, and trust we can iterate the rest.

pjhartzell replied:
Yeah, I like that much better. Thanks.

gadomski commented on src/stactools/landsat/utils.py:
```diff
@@ -98,3 +98,34 @@ def handle_antimeridian(item: Item, antimeridian_strategy: Strategy) -> None:
+        )
+
+    if item.bbox is not None:
+        item.bbox = recursive_round(list(item.bbox), precision)
```
```suggestion
        item.bbox = recursive_round(list(item.bbox), precision)
        
    return item
```

pjhartzell committed:
COORD_PRECISION --> COORDINATE_PRECISION

pjhartzell committed:
a few fixes:

- make handle_antimeridian and round_coordinates fluent
- simplify recursive_round function

pjhartzell committed:
fix round_coordinates return value

pjhartzell commented:
@gadomski Note the force push
